### PR TITLE
[SYCL] Update image accessor constructor

### PR DESCRIPTION
**Pull request by IgorKharchikov**
Changed image_accessor() constructor from "= default" to user defined to make it possible to use const references in parallel_for.

### CODE REVIEW

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// RUN: %clangxx -I %sycl_include -S -emit-llvm -fsycl -fsycl-device-only %s -o - | FileCheck %s
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:1]:
```suggestion
// RUN: %clangxx -S -fsycl -fsycl-device-only %s -o - | FileCheck %s
```
`-I` and `-emit-llvm` are not necessary when `-fsycl` and `-fsycl-device-only` are set.

What is `FileCheck` is called for?  I don't see any checks inside the code.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                            range };
+    sycl::image<2> dst_image{ sycl::image_channel_order::rgba, sycl::image_channel_type::fp32, range };
+    sycl::queue queue;
+
+    queue.submit([&](sycl::handler& handler) {
+        const auto src_accessor = src_image.get_access<sycl::float4, sycl::access::mode::read>(handler);
+        const auto dst_accessor = dst_image.get_access<sycl::float4, sycl::access::mode::write>(handler);
+        func(handler, range, src_accessor, dst_accessor);
+        });
+    return 0;
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:31]: I don't think we need all this code to test image accessor constructor. Please, reduce the test.
For instance, can we remove `main` function and leave only `func`?

File: sycl/test/check_device_code/const_ref_image_accessor.cpp (hunk 1 above)
Comment (IgorKharchikov) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: Errors in commented section for testing purposes were corrected in new commit.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp (hunk 2 above)
Comment (IgorKharchikov) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: Test file was reduced, in new commits everything is put inside main.

Comment (bader) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: "everything" has a lot of unnecessary code. Please, follow my suggestion and leave only `func` and remove `main`. To verify device code you need to add `SYCL_EXTERNAL` to `func` declaration.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// RUN: %clangxx -S -fsycl -fsycl-device-only %s
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsycl-device-only %s -o /dev/null
```
If we don't check the emitted code, `-S` is not needed as well.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp (hunk 1 above)
Comment (bader) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: Okay. I mark this as resolved, but I created another to comment to fix errors added in new commit.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp (hunk 2 above)
Comment (IgorKharchikov) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: Recommended changes were applied.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// RUN: %clangxx -fsycl -fsycl-device-only %s -o /dev/null
+
+#include <CL/sycl.hpp>
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:3]:
```suggestion
#include <sycl/sycl.hpp>
```

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+SYCL_EXTERNAL void
+func(sycl::handler &handler, const sycl::range<2> &range,
+     const sycl::accessor<sycl::float4, 2, sycl::access::mode::read,
+                          sycl::access::target::image> &dummy_accessor,
+  handler.parallel_for(range, [=](const sycl::item<2> item) {
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:9]: The code is corrupted.

**Review by bader — CHANGES_REQUESTED** [review]

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// RUN: %clangxx -fsycl -fsycl-device-only %s -o /dev/null
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:1]: BTW, you 
```suggestion
// RUN: %clangxx -fsycl -fsycl-device-only -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note,warning %s
// expected-no-diagnostics
```
This one should be even better.

**Review by bader — APPROVED** [review]
Thanks. LGTM.

**Comment by IgorKharchikov** [comment]
It appears, that new test always fails, I'm starting to investigate the problem.

**Comment by bader** [comment]
Cool. It seems we found an ICE in the SYCL front-end compiler. @intel/dpcpp-cfe-reviewers, could you take a look, please?

**Comment by IgorKharchikov** [comment]
Problem appears to be connected with SYCL_EXTERNAL, I'm currently in process of familiarizing with theory around it's appliance.

**Comment by premanandrao** [comment]
> > Problem appears to be connected with SYCL_EXTERNAL, I'm currently in process of familiarizing with theory around it's appliance.
> 
> I don't see any problems with using SYCL_EXTERNAL here. Anyway the compiler should not crash, but give an error message if SYCL_EXTERNAL usage is incorrect.

I believe @Fznamznon just fixed this with PR #5342.

**Comment by bader** [comment]
@premanandrao, thanks for letting us know.
@IgorKharchikov, please, update your branch to check if this problem is fixed.

**Comment by IgorKharchikov** [comment]
@bader, thanks, a merge was performed, but the problem still exists, though error messages are new. I'm starting to investigate them.

**Review by vladimirlaz — APPROVED** [review]

**Comment by bader** [comment]
> @bader, thanks, a merge was performed, but the problem still exists, though error messages are new. I'm starting to investigate them.

The test still uses "heavy" `parallel_for`. What is the status of the problem with using `SYCL_EXTERNAL`? Has it been resolved?

**Comment by bader** [comment]
> @bader, parallel_for was mentioned in the ticket and an ad hoc reproducer, so I decided to add it for visibility of test, as for SYCL_EXTERNAL, it appeared to be redundant here and it was causing conflicts with parallel_for.

Is parallel_for required to reproduce the issue? It doesn't look like a source of the problem. Right?

**Comment by IgorKharchikov** [comment]
@bader, in fact, parallel_for was not causing any issues, I understand your demand, the change was applied in the last commit, thanks for help in understanding of writing a succinct test.

File: sycl/test/check_device_code/const_ref_image_accessor.cpp [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+// expected-no-diagnostics
+
+#include <sycl/sycl.hpp>
+
+SYCL_EXTERNAL void
+func(const sycl::accessor<sycl::float4, 2, sycl::access::mode::read,
+                          sycl::access::target::image> &dummy_accessor) {
+  const sycl::int2 coords{0, 0};
+  auto pixel = dummy_accessor.read(coords);
+}
```
Comment (bader) [inline sycl/test/check_device_code/const_ref_image_accessor.cpp:12]: This code doesn't reproduce the problem. You need something like this:

```suggestion
SYCL_EXTERNAL void func() {
  const sycl::accessor<sycl::float4, 2, sycl::access::mode::read,
                       sycl::access::target::image>
      x;
}
```

Comment (IgorKharchikov) [reply sycl/test/check_device_code/const_ref_image_accessor.cpp]: Sorry, I now understand the problem, thanks.

**Review by vladimirlaz — APPROVED** [review]

**Pull request merged**